pydough/errors/pydough_error_builder.py

```diff
@@ -0,0 +1,344 @@
+            )
+
+    def expected_expression(self, expr: object) -> PyDoughException:
+        """
+        Creates an exception for when a QDAG expression is expected but
+        something else is found.
+        """
+        from pydough.qdag import PyDoughCollectionQDAG
+
+        if isinstance(expr, PyDoughCollectionQDAG):
+            return PyDoughQDAGException(
+                f"Expected an expression, but received a collection: {expr}"
+            )
+        else:
+            return PyDoughQDAGException(
+                f"Expected an expression, but received {expr.__class__.__name__}:  {expr}"
+            )
+
+    def type_verification_fail(
+        self, operator: "PyDoughOperator", args: list[object], message: str
+    ) -> PyDoughException:
+        """
+        Creates an exception for when type verification fails for an operator.
+
+        Args:
+            `operator`: The operator that failed type verification.
+            `args`: The arguments passed to the operator.
+            `message`: The error message explaining the typing failure.
+
+        Returns:
+            An exception indicating the type verification failure.
+        """
+        arg_strings: list[str] = [str(arg) for arg in args]
+        raise PyDoughQDAGException(
+            f"Invalid operator invocation {operator.to_string(arg_strings)!r}: {message}"
+        )
+
+    def type_inference_fail(
+        self, operator: "PyDoughOperator", args: list[object], message: str
+    ) -> PyDoughException:
+        """
+        Creates an exception for when return type inference fails for an
+        expression function operator.
+
+        Args:
+            `operator`: The operator that failed type inference.
+            `args`: The arguments passed to the operator.
+            `message`: The error message explaining the inference failure.
+
+        Returns:
+            An exception indicating the type inference failure.
+        """
+        arg_strings: list[str] = [str(arg) for arg in args]
+        raise PyDoughQDAGException(
+            f"Unable to infer the return type of operator invocation {operator.to_string(arg_strings)!r}: {message}"
+        )
+
+    def bad_columns(self, columns: object) -> PyDoughException:
+        """
+        Creates an exception for when the `columns` to `to_sql` or `to_df` is
+        not valid.
+
+        Args:
+            `columns`: The columns argument that caused the error.
+
+        Returns:
+            An exception indicating the bad `columns` argument.
+        """
+        if isinstance(columns, list):
+            for column in columns:
+                if not isinstance(column, str):
+                    return PyDoughQDAGException(
+                        f"Expected `columns` argument to be a list of strings, found {column.__class__.__name__}"
+                    )
+            return PyDoughQDAGException(
+                "Expected `columns` argument to be a non-empty list"
```

Comment:
After iterating over columns why return an exception?
e.g. If columns = ["a", "b"] this still returns exception with "Expected non-empty list"
Shouldn't this be
```
if not columns:
        return PyDoughQDAGException("Expected `columns` argument to be a non-empty list")
```

Reply:
I believe this assumes that we have bad columns which is either empty or non-string. (plus being list or dict)
I still think adding explicit check for being empty to be clear and avoid misleading like what happened to me